The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10891818881/job/30223547743) [(plain)](https://github.com/rust-lang/rust/commit/b73d7558d2469a520374d956427a30f5e53ec59e/checks/30223547743/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.642 Building wheels for collected packages: reuse
#13 2.643   Building wheel for reuse (pyproject.toml): started
#13 2.884   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.885   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.885   Stored in directory: /tmp/pip-ephem-wheel-cache-bnte0dkl/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.888 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.276 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.277 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.789 Collecting virtualenv
#13 3.789 Collecting virtualenv
#13 3.826   Downloading virtualenv-20.26.4-py3-none-any.whl (6.0 MB)
#13 3.899      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 85.3 MB/s eta 0:00:00
#13 3.957 Collecting filelock<4,>=3.12.2
#13 3.960   Downloading filelock-3.16.0-py3-none-any.whl (16 kB)
#13 3.992 Collecting platformdirs<5,>=3.9.1
#13 3.995   Downloading platformdirs-4.3.3-py3-none-any.whl (18 kB)
#13 4.013 Collecting distlib<1,>=0.3.7
#13 4.024      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 79.8 MB/s eta 0:00:00
#13 4.024      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 79.8 MB/s eta 0:00:00
#13 4.104 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.291 Successfully installed distlib-0.3.8 filelock-3.16.0 platformdirs-4.3.3 virtualenv-20.26.4
#13 DONE 4.4s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      190400 kB
DirectMap2M:     7149568 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/fd2c811d25ad5773fbd2463e058bf4edf4d44eb2/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-fd2c811d25ad5773fbd2463e058bf4edf4d44eb2-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5550 files
tidy check
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.2)
All checks passed!
checking C++ file formatting
checking C++ file formatting
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:505:32: error: code should be clang-formatted [-Wclang-format-violations]
    buffer_offset = Arg0.size()+1;
                               ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:505:33: error: code should be clang-formatted [-Wclang-format-violations]
    buffer_offset = Arg0.size()+1;
                                ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:506:40: error: code should be clang-formatted [-Wclang-format-violations]
    std::string ArgsCppStr(ArgsCstrBuff+buffer_offset, ArgsCstrBuffLen-1);
                                       ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:506:41: error: code should be clang-formatted [-Wclang-format-violations]
    std::string ArgsCppStr(ArgsCstrBuff+buffer_offset, ArgsCstrBuffLen-1);
                                        ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:506:71: error: code should be clang-formatted [-Wclang-format-violations]
    std::string ArgsCppStr(ArgsCstrBuff+buffer_offset, ArgsCstrBuffLen-1);
                                                                      ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:506:72: error: code should be clang-formatted [-Wclang-format-violations]
    std::string ArgsCppStr(ArgsCstrBuff+buffer_offset, ArgsCstrBuffLen-1);
                                                                       ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:509:34: error: code should be clang-formatted [-Wclang-format-violations]
      i = ArgsCppStr.find('\0', i+1);
                                 ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:509:35: error: code should be clang-formatted [-Wclang-format-violations]
      i = ArgsCppStr.find('\0', i+1);
                                  ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:511:32: error: code should be clang-formatted [-Wclang-format-violations]
        ArgsCppStr.replace(i, i+1, " ");
                               ^
/checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:511:33: error: code should be clang-formatted [-Wclang-format-violations]
        ArgsCppStr.replace(i, i+1, " ");


clang-format linting failed! Printing diff suggestions:
--- /checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (actual)
+++ /checkout/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (formatted)
@@ -501,15 +501,15 @@
 #if LLVM_VERSION_GE(20, 0)
     int buffer_offset = 0;
     assert(ArgsCstrBuff[ArgsCstrBuffLen - 1] == '\0');
     std::string Arg0(ArgsCstrBuff);
-    buffer_offset = Arg0.size()+1;
-    std::string ArgsCppStr(ArgsCstrBuff+buffer_offset, ArgsCstrBuffLen-1);
+    buffer_offset = Arg0.size() + 1;
+    std::string ArgsCppStr(ArgsCstrBuff + buffer_offset, ArgsCstrBuffLen - 1);
     auto i = 0;
     while (i != std::string::npos) {
-      i = ArgsCppStr.find('\0', i+1);
+      i = ArgsCppStr.find('\0', i + 1);
       if (i != std::string::npos)
-        ArgsCppStr.replace(i, i+1, " ");
+        ArgsCppStr.replace(i, i + 1, " ");
     }
     Options.MCOptions.Argv0 = std::string(Arg0);
     Options.MCOptions.CommandlineArgs = ArgsCppStr;
 #else

tidy error: checks with external tool 'clang-format' failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Mon Sep 16 20:55:27 UTC 2024
  network time: Mon, 16 Sep 2024 20:55:27 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
+++ b/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
@@ -503,7 +503,8 @@ extern "C" LLVMTargetMachineRef LLVMRustCreateTargetMachine(
     assert(ArgsCstrBuff[ArgsCstrBuffLen - 1] == '\0');
     auto Arg0 = std::string(ArgsCstrBuff);
     buffer_offset = Arg0.size() + 1;
-    auto ArgsCppStr = std::string(ArgsCstrBuff + buffer_offset, ArgsCstrBuffLen - 1);
+    auto ArgsCppStr =
+        std::string(ArgsCstrBuff + buffer_offset, ArgsCstrBuffLen - 1);
     auto i = 0;
     while (i != std::string::npos) {
       i = ArgsCppStr.find('\0', i + 1);

--- a/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
+++ b/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
@@ -498,6 +498,21 @@ extern "C" LLVMTargetMachineRef LLVMRustCreateTargetMachine(
   Options.EmitStackSizeSection = EmitStackSizeSection;
 
   if (ArgsCstrBuff != nullptr) {
+#if LLVM_VERSION_GE(20, 0)
+    int buffer_offset = 0;
+    assert(ArgsCstrBuff[ArgsCstrBuffLen - 1] == '\0');
+    auto Arg0 = std::string(ArgsCstrBuff);
+    buffer_offset = Arg0.size() + 1;
+    auto ArgsCppStr = std::string(ArgsCstrBuff + buffer_offset, ArgsCstrBuffLen - 1);
+    auto i = 0;
+    while (i != std::string::npos) {
+      i = ArgsCppStr.find('\0', i + 1);
+      if (i != std::string::npos)
+        ArgsCppStr.replace(i, i + 1, " ");
+    }
+    Options.MCOptions.Argv0 = Arg0;
+    Options.MCOptions.CommandlineArgs = ArgsCppStr;
+#else
     int buffer_offset = 0;
     assert(ArgsCstrBuff[ArgsCstrBuffLen - 1] == '\0');
 
@@ -523,6 +538,7 @@ extern "C" LLVMTargetMachineRef LLVMRustCreateTargetMachine(
     Options.MCOptions.Argv0 = arg0;
     Options.MCOptions.CommandLineArgs =
         llvm::ArrayRef<std::string>(cmd_arg_strings, num_cmd_arg_strings);
+#endif
   }
 
   TargetMachine *TM = TheTarget->createTargetMachine(
@@ -531,10 +547,11 @@ extern "C" LLVMTargetMachineRef LLVMRustCreateTargetMachine(
 }
 
 extern "C" void LLVMRustDisposeTargetMachine(LLVMTargetMachineRef TM) {
-
+#if LLVM_VERSION_LT(20, 0)
   MCTargetOptions &MCOptions = unwrap(TM)->Options.MCOptions;
   delete[] MCOptions.Argv0;
   delete[] MCOptions.CommandLineArgs.data();
+#endif
 
   delete unwrap(TM);
 }

